File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/type/Type.scala
package io.github.chaosunity.casc.parsing.`type`

trait Type {
    def name: String

    def classType: Class[_]

    def descriptor: String

    def internalName: String

    def loadVariableOpcode: Int

    def storeVariableOpcode: Int

    def returnOpcode: Int

    def addOpcode: Int

    def subtractOpcode: Int

    def multiplyOpcode: Int

    def divideOpcode: Int
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/type/Type.kt
package io.github.chaosunity.casc.parsing.type

import jdk.internal.org.objectweb.asm.Opcodes

interface Type {
    val typeName: String
    fun classType(): Class<*>?
    val descriptor: String
    val internalName: String

    val loadVariableOpcode: Int
    val storeVariableOpcode: Int
    val returnOpcode: Int
    val addOpcode: Int
    val subtractOpcode: Int
    val multiplyOpcode: Int
    val divideOpcode: Int

    fun isInt(): Boolean =
        this == BuiltInType.INT

    fun isLong(): Boolean =
        this == BuiltInType.LONG

    fun isFloat(): Boolean =
        this == BuiltInType.FLOAT

    fun isDouble(): Boolean =
        this == BuiltInType.DOUBLE

    fun isBool(): Boolean =
        this == BuiltInType.BOOLEAN

    fun isString(): Boolean =
        this == BuiltInType.STRING

    fun isBuiltInType() =
        BuiltInType.values().any { it == this }
}